[wx] Fix expiration controls

nobugshere opened the pull request:
Resolves #935
Resolves #1149
Reinstates FR890 (Recurring checkbox on by default when interval is selected)

This is a major re-work of the password expiration controls.  It took me a while to work out all the corner cases so, please, give it a good shake down.
- Fixed numerous bugs in the control management and cases of false or missed change detection.
- Fixed multiple issues when adding a new entry
- Use the default interval set in Options | Password History tab.
- Recurring checkbox is set by default when the interval radio button is selected
- Truncate Date/Time values to use only the date, this fixed a number of comparison problems
    (The Date picker control handles time differently on OSX vs. GTK)
- Added “(Expired)” or “(Expires in %d days)” to the “Original Value:” text
- Added a description of the records to the formatV3.txt document

Tested on:
macOS M1Max Sonoma 14.6,   Xcode 15.4, wx 3.2.4 and 3.2.5 (Development and most testing)
Ubuntu 24.04 ARM64 VMware, Gnome/Wayland, gcc 13.2.0, wx 3.2.4, GTK 3.24.41
Fedora 40 ARM64 VMware,    Xfce/X11,      gcc 14.1.1, wx 3.2.4, GTK 3.24.43
Fedora 40 ARM64 VMware,    KDE/Wayland,   gcc 14.1.1, wx 3.2.4, GTK 3.24.43
Fedora 38 ARM64 VMware,    Xfce/X11,      gcc 13.2.1, wx 3.0.5, GTK 3.24.41

The only layout change is to add “(Expired)” or “(Expires in %d days)” to the Original Value text:
![Screenshot markup](https://github.com/user-attachments/assets/02d7c77d-305c-472f-8c1a-82e6875aa6a6)

dskrvk commented on src/ui/wxWidgets/AddEditPropSheetDlg.h:
```diff
@@ -388,10 +392,12 @@ class AddEditPropSheetDlg : public wxPropertySheetDialog
   bool m_Recurring;
-  wxString m_ExpirationTime;
+  bool m_OriginalRecurring;
+  time_t m_OriginalDayttt;
```
Not sure what the third "t" stands for, but for consistency with the other time_t member:
```suggestion
  time_t m_tttOriginalDay;
```

nobugshere replied:
Well...I'm going for consistency with the other m_Original's :)

dskrvk commented on src/ui/wxWidgets/AddEditPropSheetDlg.cpp:
```diff
@@ -2656,16 +2668,40 @@ void AddEditPropSheetDlg::OnExpRadiobuttonSelected( wxCommandEvent& evt )
   bool Never = (evt.GetEventObject() == m_DatesTimesNeverExpireCtrl);
```
Nit: redundant parentheses on these two lines

nobugshere replied:
Removed.

dskrvk commented on src/ui/wxWidgets/AddEditPropSheetDlg.cpp:
```diff
@@ -2656,16 +2668,40 @@ void AddEditPropSheetDlg::OnExpRadiobuttonSelected( wxCommandEvent& evt )
+    m_tttExpirationTime = xdt.GetTicks();
+    m_Recurring = true;
   }
+  // In add mode, the item data is not prepared for this, fields will go blank.
```
This seems like a tight coupling between this event handler and the control, in that we're making assumptions about another control's behavior. Instead, could we still send the data, but implement proper handling logic in the control? I don't know how we're populating the field there, so this may not make sense.

nobugshere replied:
I think the problem is that the function that initializes the dialog controls (it was called UpdateExpTimes, I changed it to InitializeExpTimes because it's only called once) finishes by calling OnExpRadiobuttonSelected() to complete the initialization.  It worked okay when editing an existing entry, but caused problems when filling-in a new entry.  I thought it was strange to call an event handler like that, but I didn't change it, I just tried to make it work.  

I'm going to re-think that decision and see if I can refactor that code for a cleaner separation.

dskrvk commented on src/ui/wxWidgets/AddEditPropSheetDlg.cpp:
```diff
@@ -2656,16 +2668,40 @@ void AddEditPropSheetDlg::OnExpRadiobuttonSelected( wxCommandEvent& evt )
-    m_tttExpirationTime = time_t(0);
-    m_ExpirationTimeInterval = 90;
-    wxDateTime xdt(wxDateTime::Now());
+    //m_tttExpirationTime = time_t(0);
```
Remove this line?

nobugshere replied:
Removed.

dskrvk commented on src/ui/wxWidgets/AddEditPropSheetDlg.cpp:
```diff
@@ -2656,16 +2668,40 @@ void AddEditPropSheetDlg::OnExpRadiobuttonSelected( wxCommandEvent& evt )
+    // which is interpreted as "Never".  But, we need to use the given interval to prevent
+    // a false change detection when the entry is opened and closed.  If an actual change is made,
+    // the entry will be re-written correctly.
+    if (!m_ExpirationTimeInterval) { // Compatibility
```
To confirm, there's no "never" setting anymore - if the user hasn't specified any expiration settings, we'll always use the default interval - is that right?
E.g. if the user selected "never", why do we set the expiry date? Wouldn't that cause the entry to be marked expired on that date, contrary to what the user wanted?

And for additional context, is the problem basically that there was a bug which caused a nonsensical combination of settings to get stored (interval but no date), so we're doing a one-time migration to bring the stored settings to a valid state?

nobugshere replied:
> To confirm, there's no "never" setting anymore - if the user hasn't specified any expiration settings, we'll always use the default interval - is that right?

There never was a "never" parameter written to the file, it's the absence of the date and interval fields.  Upon initialization of the dialog, my intention is to pre-populate the controls with the default interval and the corresponding date in case the user wants to use them, but the "Never" radio button will be selected.

> E.g. if the user selected "never", why do we set the expiry date? Wouldn't that cause the entry to be marked expired on that date, contrary to what the user wanted?

The control values won't matter if the "never" radio button is selected, neither field is written to the file.  Also, Never is still the default for a new entry.

> And for additional context, is the problem basically that there was a bug which caused a nonsensical combination of settings to get stored (interval but no date)

Exactly, the current version is doing that when Never is selected, so there could be lots of files with that condition.  

> so we're doing a one-time migration to bring the stored settings to a valid state?

No, I didn't want to make mysterious changes to the file.  I accept that condition as Never. (As does the current release.) But, if the user makes an expiry change, it will be written correctly.  One important goal to me is that the user can open the edit dialog and cancel without any changes and not have a false change detection pop-up or write changes to the file.
I did use that interval field to pre-populate the control.  I'm starting to re-consider that, though.  Maybe I should use the default value and completely ignore that field.

Any additional thoughts are welcome.

dskrvk commented on src/ui/wxWidgets/AddEditPropSheetDlg.cpp:
```diff
@@ -2656,16 +2668,40 @@ void AddEditPropSheetDlg::OnExpRadiobuttonSelected( wxCommandEvent& evt )
     xdt += wxDateSpan(0, 0, 0, m_ExpirationTimeInterval);
     m_DatesTimesExpiryDateCtrl->SetValue(xdt);
+    m_tttExpirationTime = xdt.GetTicks();
+    m_Recurring = (m_Type == SheetType::ADD) ? true : false;
```
```suggestion
    m_Recurring = m_Type == SheetType::ADD;
```

nobugshere replied:
Uh...I must have been tired :)  Fixed.

nobugshere commented:
I removed the connection between the initialization routine and the radio button handler.  I also changed the special case to ignore the improperly written interval and use the user-set default instead.  (The entry won't be re-written unless the user makes a change.)   I think this version much cleaner, though there are still a lot of possible variations, especially changing from one mode to another.  I tried to handle everything I could think of :)  Please, have another look.

dskrvk commented on src/ui/wxWidgets/AddEditPropSheetDlg.cpp:
```diff
@@ -1347,52 +1343,95 @@ static struct {short pv; wxString name;}
+  // an interval but no date, which is interpreted as "Never".  We are going to ignore
+  // it and use the user-set default.  If an expiry change is made, the entry will
+  // be re-written correctly.
+  int defInterval = PWSprefs::GetInstance()->GetPref(PWSprefs::DefaultExpiryDays);
```
```suggestion
  int defaultInterval = PWSprefs::GetInstance()->GetPref(PWSprefs::DefaultExpiryDays);
```

nobugshere replied:
Done

dskrvk commented on src/ui/wxWidgets/AddEditPropSheetDlg.h:
```diff
@@ -288,8 +288,14 @@ class AddEditPropSheetDlg : public wxPropertySheetDialog
   void SetXTime(wxObject *src); // sync controls + controls -> entry
+
+  // Today returns time part == 0
+  wxDateTime TodayPlusInterval() { return wxDateTime::Today().Add(wxDateSpan(0, 0, 0, m_ExpirationTimeInterval)); };
```
Stylistic preference, but for simple functions like this it's better IMO to not have side effects and instead take all inputs as parameters. It's just not obvious from the prototype what interval we're adding. Also, the function would be more reusable with a parameter for interval.

nobugshere replied:
Good point.  Changed.

dskrvk commented on src/ui/wxWidgets/AddEditPropSheetDlg.cpp:
```diff
@@ -1347,52 +1343,95 @@ static struct {short pv; wxString name;}
+  int defInterval = PWSprefs::GetInstance()->GetPref(PWSprefs::DefaultExpiryDays);
 
-  wxCommandEvent dummy;
   wxDateTime exp;
```
```suggestion
  wxDateTime expiryDate;
```
(I had to scroll back to this line a few times to remind myself what this variable contains)

nobugshere replied:
Done

dskrvk commented on src/ui/wxWidgets/AddEditPropSheetDlg.cpp:
```diff
@@ -1347,52 +1343,95 @@ static struct {short pv; wxString name;}
+  // time part set to 0, and Today() and Now() both use the local time zone.
+  m_DatesTimesExpiryDateCtrl->SetValue(exp);
+  m_tttExpirationTime = exp.GetTicks();
+  m_Recurring = true;
```
Why is this set to true unconditionally?

nobugshere replied:
This is just setting the default state of the recurring checkbox when the dialog is initialized.  All three of the conditions in the preceding if/else blocks would set it to true, so I did it in one place to highlight it as a common default.   I've added a comment.  (Note: this is consistent with the Windows version.)

dskrvk commented on src/ui/wxWidgets/AddEditPropSheetDlg.cpp:
```diff
@@ -1347,52 +1343,95 @@ static struct {short pv; wxString name;}
+
+  if (m_OriginalDayttt) {
+    wxString rstr;
+    int interval = IntervalFromDate(wxDateTime(m_OriginalDayttt));
```
```suggestion
    int interval = IntervalFromDate(exp);
```
or is it crucial that we include the time part here?

nobugshere replied:
This is building a string to represent the number of days until expiration, so it needs the original date.  At this point, exp (expiryDate) may have changed.

nobugshere commented:
While working on the third round, I realized m_tttExpirationTime is always set to match the value of the m_DatesTimesExpiryDateCtrl date picker control, so it's not needed.

ronys commented:
I really appreciate the work you put in here, as the code is (very) tricky to get right. In fact, going back to the Windows version to compare behavior, I still see problems there...

A couple of issues that seem worth addressing:

1. Create a new entry, specify an expiration date. The "Never" button is still selected even though I've specified a date. So is the date in effect, or do I also have to explicitly click on the "On" radio button?
2. Likewise if I sent an interval.
3. Even more confusing: I set a date, then changed the interval, radio button left on "Never". Close and re-open entry, "On" is selected.

Also: setting a new entry by date, the interval track the date, which is nice, but closing and then opening the entry, the interval reverts to the default, which is confusing.

Finally, I think the text for original value when an interval's selected is awkward, e.g., "08/23/2024 (expires in 12 days) (every 12 days)". Perhaps the following's a bit clearer:
- When the date's specified: "08/23/2024"
- When an interval's specified: "12 day from last change (expiring on 08/23/2024)"
- When a recurring interval's specified: "Every 12 days from last change (next expiration on 08/23/2024)"

nobugshere commented:
Hi @ronys,
Thanks for the feedback.  I hope it's getting close :)

First of all, do you agree with the way I described the feature in the format docs?  Much of it was inferred.  I need to get that clear in my head first :)

- 1 and 2
My intention is for the radio buttons to determine which mode is being set.  Which, in turn, determines which other controls are meaningful.  The Windows version actually enables and disables the controls depending on which radio button is chosen. There is code in the WX version to do the same thing, but for some reason, it isn't working.  Issue #1220 is related.  I suspect it's a problem in wxWidgets, but so far, my attempt to make a simplified test case has failed to reproduce the problem.

- 3
I cannot reproduce this scenario on Mac, Ubuntu or Fedora.  Can you provide more info on the platform you're using?  And, was this a new entry or editing an existing entry?

> Also: setting a new entry by date, the interval track the date, which is nice, but closing and then opening the entry, the interval reverts to the default, which is confusing.

I went back and forth on this a few times; I don't remember why I left it this way.  I can change it to initialize the interval field based on the number of days until the specified date, if that makes more sense.  In that case, I would only use the default for a new entry or an existing entry that is currently set to Never.

> Finally, I think the text for original value when an interval's selected is awkward...

The static text "Original value" seems awkward too.  Would something like "Current Setting" be better?

This implementation doesn't record a case of an interval that is not recurring.  That's just recorded as a specific date.  So, how about something like this?  Perhaps, rendering expired messages in red?

- Specific date:
Expires in X days (\<date\>) -or- Expired on (\<date\>)
- Recurring interval:
Every 12 days from last change (Next expiration on \<date\>) -or- 
Every 12 days from last change (Expired on \<date\>)
- Never
Never Expires

ronys commented:
Hi @nobugshere,

Thanks for pointing to the format description - I was intent on checking the functionality, and missed that change. To the point: I think the only way case 3 (only interval field defined) occurred was if the user set an interval and the disabled it by selecting the "Never" radio button. This may have changed in later versions to just not recording it in this case. The semantics are indeed as you described: the value is ignored.

I'm testing on Debian 12 (bookworm), which uses wxWidgets 3.2.2. Scenario 3 happens both in existing and new entries.

Re dates and interval: The problem is the inconsistent behavior. I like the tracking, that is, setting the interval to the number of remaining days, but if it's easier to leave the default, then that's fine, as long as that's always what's displayed when the date's set.

Expiration texts: "Current Setting" is fine, as well as your proposed texts, except for the redundant parenthesis in "Expired on (<date>)" - should be just "Expired on <date>"

nobugshere commented:
> I'm testing on Debian 12 (bookworm), which uses wxWidgets 3.2.2. Scenario 3 happens both in existing and new entries.

I'll look into setting-up a Debian 12 ARM VM.  Which desktop environment do you use?

Just to be clear:
1. Open an entry or create a new one
2. Don't touch the radio buttons (Never is selected)
3. Set a date (via text entry or date chooser?)
4. Change the interval (via text entry or spin control?)
5. click OK
6. Re-open the entry and the "on" radio button is selected.

nobugshere commented:
@ronys 
Made changes as discussed above.  Please, see what you think.
1. When opening an entry with a specific date set, initialize the interval to "days until expiration" instead of the default value.
2. Removed some -W options that are not needed after @huven PR's.
3. I still cannot reproduce the error you reported, not even on Debian 12.6 ARM, wx3.2.2, and Xfce desktop.
4. Re-worked the expiration text.  The text for recurring intervals was too long so I put it on two lines.  I also turn the text red when the password is expired.

![Screenshot 2024-08-15 at 11 30 18 AM](https://github.com/user-attachments/assets/1736320f-a7fa-420d-98d4-58f4075c09a1)

ronys commented:
The attached clip illustrates the problem I'm seeing.
[clip.zip](https://github.com/user-attachments/files/16643936/clip.zip)
It looks like changes to the expiry aren't detected/saved.
Specifically:
- the 'recurring' checkbox is always set
- changing from an absolute date to interval isn't persistent.

Also, I think the relevant radio button should be set when the control to its right gets focus.

Database is also attached, password "Sample123".
[safe.zip](https://github.com/user-attachments/files/16643967/safe.zip)

nobugshere commented:
@ronys,
I finally figured-out why the controls could not be disabled (grayed-out) corresponding to the selected radio button.  So, now I can offer you a choice:

1. Enable/disable the controls depending on the radio button selected (consistent with the Windows version) (I would still strive to keep the date and interval days in sync, just because.)
2. Continue down the current path where all controls are active and the RB follows focus.

IMHO, No.1 is probably simpler overall with fewer chances of corner cases and, maybe, less confusing.  But, I can go either way.

ronys commented:
Excellent!
I agree: Consistency with the Windows version + date-interval sync sounds clearest, even if it means an extra mouse click for the user.

nobugshere commented:
@ronys,
Once more into the breach....

Resolves #1220 

Now the controls are enabled/disabled depending on which radio button is selected.  Commit 5dc399f is the fix for that.

Some things you might notice:
1. When a specific date is set, the interval is initialized to the days until expiration except, if it's already expired, the interval is initialized to the default value.
2. The Recurring checkbox is checked the first time the "In" radio button is selected, whether by initialization or by the user.  After that, it maintains its state until the dialog is closed.

I think this covers the condition in your video by virtu of the fact the controls are grayed-out.

ronys commented:
Looks good to me.
@dskrvk - any comments?
